### ISSUE TITLE
[BUG] /delchat fails to delete current chat

### ISSUE THREAD

**Issue by Technologicat** [SillyTavern/SillyTavern#1772]
### Environment

Self-Hosted (Bare Metal)

### System

Firefox 121.0, Linux Mint 21.1

### Version

SillyTavern 1.11.3 'staging' (3cf01e76)

### Desktop Information

Node.js: v20.10.0
API: oobabooga (self-hosted / bare metal), with the OAI compatible endpoints
Branch: staging
Model: [dolphin-2.6-mistral-7b-dpo-laser.Q5_K_M.gguf](https://huggingface.co/TheBloke/dolphin-2.6-mistral-7B-dpo-laser-GGUF)

### Describe the problem

**Steps to reproduce**:

- Make a new chat
- Enter some message, send it to AI
- Try to delete the current chat file by issuing the `/delchat` slash command

**Actual result**:

Nothing happens.

**Expected result**:

The current chat file is deleted, and the latest previous chat file (with the same character) is automatically opened (similar to how it works if you delete the current chat file in *Manage chat files*).

### Additional info

Checked `console.log`, there is nothing useful.

I'd like to use this and `/buttons` to create a Quick Reply button to quickly delete the current chat, with confirmation.

### Please tick the boxes

- [X] You have explained the issue clearly, and included all relevant info
- [X] You've checked that this [issue hasn't already been raised](https://github.com/SillyTavern/SillyTavern/issues?q=is%3Aissue)
- [X] You've checked the [docs](https://docs.sillytavern.app/) ![important](https://img.shields.io/badge/Important!-F6094E)

**EDIT**: *Fixed a typo.*

**Comment by Cohee1207** [comment]
I wasn't able to reproduce. `/delchat` works when used in command, but not when QR reply?

**Comment by Technologicat** [comment]
For me, doesn't work even directly as a command.

Nothing is printed in the server terminal window, and here's all that appears in the client console:

![image](https://github.com/SillyTavern/SillyTavern/assets/16972251/77a6ae2c-d191-440d-8745-d9b8a48352c3)

I'll try to trace it and see what it's about. I'll get back to you later on this.

**Comment by Technologicat** [comment]
Ok, traced it. When in doubt, add copious debug logging:

```javascript
async function doDeleteChat() {
    console.debug('doDeleteChat entered');
    $('#option_select_chat').trigger('click', { fromSlashCommand: true });
    await delay(100);
    console.debug('Highlighted chat block:');
    const highlightedChatBlock = $('.select_chat_block[highlight=\'true\']');
    console.debug(highlightedChatBlock);
    console.debug('The parent of the highlighted chat block:');
    const parentOfHighlightedChatBlock = highlightedChatBlock.parent();
    console.debug(parentOfHighlightedChatBlock);
    let currentChatDeleteButton = parentOfHighlightedChatBlock.find('.PastChat_cross');
    console.debug('Delete button for current chat:');
    console.debug(currentChatDeleteButton);
    $(currentChatDeleteButton).trigger('click', { fromSlashCommand: true });
    await delay(1);
    $('#dialogue_popup_ok').trigger('click');
    //200 delay needed let the past chat view reshow first
    await delay(200);
    $('#select_chat_cross').trigger('click');
    console.debug('doDeleteChat done');
}
```

and this is what I get (Firefox 121.0):

```
doDeleteChat entered
Highlighted chat block:
Object { length: 0, prevObject: {…} }
The parent of the highlighted chat block:
Object { length: 0, prevObject: {…} }
Delete button for current chat:
Object { length: 0, prevObject: {…} }
doDeleteChat done
```

On Chromium (120.0.6099.109 (Official Build) for Linux Mint (64-bit)):

```
doDeleteChat entered
Highlighted chat block:
S.fn.init
  length: 0
  prevObject: S.fn.init {0: document, length: 1}
  [[Prototype]]: Object
The parent of the highlighted chat block:
S.fn.init
  length: 0
  prevObject: S.fn.init {length: 0, prevObject: S.fn.init}
  [[Prototype]]: Object
Delete button for current chat:
S.fn.init
  length: 0
  prevObject: S.fn.init {length: 0, prevObject: S.fn.init}
  [[Prototype]]: Object
doDeleteChat done
```

So it's not finding the UI controls.

I'll think a bit on this, but ideas are welcome. :)

**Comment by Technologicat** [comment]
If I log `$('.select_chat_block')`, the result is:

```
Object { 0: div.select_chat_block.wide100p.flex-container, length: 1, prevObject: {…} }
​  0: <div class="select_chat_block  wide100p flex-container" file_name="">
​    length: 1
​  prevObject: Object { 0: HTMLDocument http://127.0.0.1:8000/, length: 1 }
​  <prototype>: Object { jquery: "3.5.1", constructor: S(e, t), length: 0, … }
```

which doesn't seem to have a `highlight` attr.

**Comment by Technologicat** [comment]
Figured it out!

I have about 130 chat files with my primary AI character. The delay of 100ms in `doDeleteChat` isn't nearly long enough to open the *Manage chat files* view.

Hence the list never gets populated, and the rest of the lookups fail.

Tested by changing the delay to 2000ms. While that makes the UX intolerable, then the deletion works.

I think the right solution would be to wait until the view has opened.

Some kind of event?

Or maybe an async function that internally awaits small delays until the view has fully opened, and then returns?

**Comment by Technologicat** [comment]
Easy way to fix this: looking at what the click handler (https://github.com/SillyTavern/SillyTavern/blob/staging/public/script.js#L8619) does, we can change `doDeleteChat` as follows:

```javascript
async function doDeleteChat() {
    await displayPastChats();
    let currentChatDeleteButton = $('.select_chat_block[highlight=\'true\']').parent().find('.PastChat_cross');
    $(currentChatDeleteButton).trigger('click', { fromSlashCommand: true });
    await delay(1);
    $('#dialogue_popup_ok').trigger('click');
    //200 delay needed let the past chat view reshow first
    await delay(200);
    $('#select_chat_cross').trigger('click');
}
```

Tested, works for me.

The *Manage chat files* view doesn't close automatically, though - same reason, the 200ms isn't enough.

**EDIT**: And we should be able to fix that by disabling the popup - adding `customData` and sending in a `fromSlashCommand` parameter.

Also, this approach lets us avoid the shadow disabling hack in the click handler. Arguably cleaner to directly call `displayPastChats`, too.

I'll post a PR.

**EDIT**: No, let's not disable the popup, it has important logic. But we can take in a `customData` in the click handler, and simply not re-show the history if the click came from the slash command. This is what's implemented in the PR.

**Comment by Technologicat** [comment]
Fix posted as #1799.

**Comment by Technologicat** [comment]
Resolved by #1799.
